cdds/cdds/extract/validate.py

```diff
@@ -168,3 +157,91 @@ def validate_directory_netcdf(path, validation_result):
                 validation_result.add_file_content_error(error)
+
+
+def get_stash_fields(path: str, validation_result: StreamValidationResult) -> dict[str, dict[int, int]]:
+    """ Validates if pp files in a given location contain all required
+    stash codes,
+
+    :param path: Path of interest.
+    :type path: str
+    :param validation_result: Holds errors and results of validation.
+    :type validation_result: StreamValidationResult
+    :return:
+    :rtype: dict[in, int]
+    """
+    stash_in_file = {}
+    pp_files = [file for file in os.listdir(path) if file.endswith('.pp')]
+
+    if pp_files:
+        for pp_file in sorted(pp_files):
+            stash = get_stash_from_pp(os.path.join(path, pp_file))
+            if stash is None:
+                validation_result.add_file_content_error(
+                    FileContentError(os.path.join(path, pp_file), "unreadable file"))
+            stash_in_file[pp_file] = stash
+    return stash_in_file
+
+
+def check_expected_stash(
+    stash_in_file: dict[str, dict[int, int]],
+    validation_result: StreamValidationResult,
+    path: str,
+    expected_stash: set[int],
+):
+    """ Checks that all the expected stash codes are found in each file.
+
+    :param stash_in_file: Stash entries in files.
+    :type stash_in_file: dict[str, dict[int, int]]
+    :param validation_result: Validation results for a given stream.
+    :type validation_result: StreamValidationResult
+    :param path: Path to files.
+    :type path: str
+    :param expected_stash: The set of expected stash codes.
+    :type expected_stash: set[int]
+    """
+
+    for file, stash in stash_in_file.items():
+        if expected_stash.difference(set(stash.keys())):
+            stash_diff = expected_stash.difference(set(stash.keys()))
+            if stash_diff:
+                error = StashError(os.path.join(path, file), "STASH errors")
+                for diff in stash_diff:
+                    error.add_stash_error(diff)
+                validation_result.add_file_content_error(error)
+
+
+def check_consistent_stash(
+    stash_in_file: dict,
+    validation_result: StreamValidationResult,
+    path: str,
+    file_frequency: str
+):
+    """ Checks that the number of stash entries is consistent across all files.
+
+    :param stash_in_file: Stash entries in files.
+    :type stash_in_file: dict
+    :param validation_result: Validation results for a given stream.
+    :type validation_result: StreamValidationResult
+    :param path: Path to files.
+    :type path: str
+    """
+
+    if Calendar.default().mode == "gregorian" and file_frequency in ["monthly", "seasonal"]:
+        logger = logging.getLogger(__name__)
+        logger.info("Skipping stash consistency check due to gregorian calendar.")
+        return None
+
+    referencedict = ()
+    for file, stash in stash_in_file.items():
+        if referencedict == ():
+            referencedict: tuple[str, dict] = (file, stash)
```

Comment:
The following sorts the type checking
```suggestion
    referencedict: tuple[str, dict] = (None, None)
    for file, stash in stash_in_file.items():
        if referencedict == (None, None):
            referencedict = (file, stash)
```